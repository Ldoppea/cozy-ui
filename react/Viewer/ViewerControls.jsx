import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import cx from 'classnames'
import Hammer from 'hammerjs'
import { withStyles } from '@material-ui/core/styles'

import { toolbarPropsPropType } from './index'
import { infoWidth } from './InformationPanel'
import Toolbar from './Toolbar'
import Navigation from './Navigation'

import styles from './styles.styl'

const ACTIONS_HIDE_DELAY = 3000

const customStyles = () => ({
  viewerControlsWithInfo: {
    width: `calc(100% - ${infoWidth}) !important`
  }
})

class ViewerControls extends Component {
  state = {
    hidden: false,
    gestures: null
  }

  _mounted = false

  showControls = () => {
    if (this._mounted) {
      this.setState({ hidden: false })
      this.hideAfterDelay()
      document.removeEventListener('mousemove', this.showControls)
      document.addEventListener('mousemove', this.hideAfterDelay)
    }
  }

  hideControls = () => {
    if (this._mounted) {
      this.setState({ hidden: true })
      document.removeEventListener('mousemove', this.hideAfterDelay)
      document.addEventListener('mousemove', this.showControls)
    }
  }

  onTap = () => {
    if (this.state.hidden) this.showControls()
    else this.hideAfterDelay()
  }

  hideAfterDelay = () => {
    clearTimeout(this.hideTimeout)
    this.hideTimeout = setTimeout(() => {
      this.hideControls()
    }, ACTIONS_HIDE_DELAY)
  }

  onSwipe = e => {
    if (e.direction === Hammer.DIRECTION_LEFT) this.props.onNext()
    else if (e.direction === Hammer.DIRECTION_RIGHT) this.props.onPrevious()
  }

  componentDidMount() {
    this._mounted = true
    clearTimeout(this.hideTimeout)
    this.hideAfterDelay()
    //eslint-disable-next-line react/no-find-dom-node
    const gestures = new Hammer(ReactDOM.findDOMNode(this.wrapped))
    gestures.on('swipe', this.onSwipe)
    gestures.on('tap', this.onTap)
    this.setState({ gestures })
  }

  componentWillUnmount() {
    this._mounted = false
    clearTimeout(this.hideTimeout)
    if (this.state.gestures) this.state.gestures.destroy()
  }

  renderChildren(children) {
    if (!children) return null
    return React.cloneElement(children, {
      gestures: this.state.gestures,
      gesturesRef: this.wrapped,
      onSwipe: this.onSwipe
    })
  }

  render() {
    const {
      currentFile,
      onClose,
      hasPrevious,
      hasNext,
      onPrevious,
      onNext,
      isMobile,
      expanded,
      toolbarProps,
      showNavigation,
      showInfoPanel,
      children,
      isMobileApp,
      classes
    } = this.props
    const { showToolbar, showClose } = toolbarProps
    const { hidden } = this.state

    return (
      <div
        className={cx(styles['viewer-controls'], {
          [styles['viewer-controls--expanded']]: expanded,
          [classes.viewerControlsWithInfo]: showInfoPanel
        })}
        ref={wrapped => {
          this.wrapped = wrapped
        }}
      >
        {showToolbar && (
          <Toolbar
            currentFile={currentFile}
            onClose={showClose && onClose}
            isMobileApp={isMobileApp}
            onMouseEnter={this.showControls}
            onMouseLeave={this.hideControls}
            isMobile={isMobile}
          />
        )}
        {showNavigation && !isMobile && hasPrevious && (
          <Navigation
            className={styles['viewer-nav--previous']}
            hidden={hidden}
            onMouseEnter={this.showControls}
            onMouseLeave={this.hideControls}
            onClick={onPrevious}
          />
        )}
        {this.renderChildren(children)}
        {showNavigation && !isMobile && hasNext && (
          <Navigation
            className={styles['viewer-nav--next']}
            hidden={hidden}
            onMouseEnter={this.showControls}
            onMouseLeave={this.hideControls}
            onClick={onNext}
          />
        )}
      </div>
    )
  }
}

ViewerControls.propTypes = {
  currentFile: PropTypes.object.isRequired,
  onClose: PropTypes.func.isRequired,
  hasPrevious: PropTypes.bool.isRequired,
  hasNext: PropTypes.bool.isRequired,
  onPrevious: PropTypes.func.isRequired,
  onNext: PropTypes.func.isRequired,
  isMobile: PropTypes.bool.isRequired,
  expanded: PropTypes.bool.isRequired,
  toolbarProps: PropTypes.shape(toolbarPropsPropType),
  showNavigation: PropTypes.bool.isRequired,
  isMobileApp: PropTypes.bool.isRequired,
  showInfoPanel: PropTypes.bool
}

export default withStyles(customStyles)(ViewerControls)
