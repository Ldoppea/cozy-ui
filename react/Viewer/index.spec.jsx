import React from 'react'
import { Viewer } from './index.jsx'
import ViewerControls from './ViewerControls'

function flushPromises() {
  return new Promise(resolve => setImmediate(resolve))
}

describe('Viewer', () => {
  const mockClient = {
    collection: jest.fn(() => ({
      getDownloadLinkById: jest.fn(id => `https://testurl.com/${id}`)
    }))
  }
  const shallowRenderOptions = {
    context: {
      client: mockClient
    }
  }

  const getViewerComponent = async component => {
    const wrapper = component.find('ViewerWrapper').dive()
    const withFileUrlWrapper = wrapper.find('withFileUrlClass')

    if (withFileUrlWrapper.length > 0) {
      const viewer = withFileUrlWrapper.dive()
      await flushPromises()
      return viewer
    } else {
      const viewer = wrapper.find(ViewerControls.displayName).childAt(0)
      return viewer
    }
  }

  it('should render the various viewers', async () => {
    const files = [
      {
        _id: 'audiofileid',
        class: 'audio',
        mime: 'audio/mp3',
        name: 'sample.mp3'
      },
      {
        _id: 'imagefileid',
        class: 'image',
        mime: 'image/jpg',
        name: 'sample.jpg'
      },
      {
        _id: 'pdffileid',
        class: 'pdf',
        mime: 'application/pdf',
        name: 'sample.pdf'
      },
      {
        _id: 'textfileid',
        class: 'text',
        mime: 'text/plain',
        name: 'sample.txt'
      },
      {
        _id: 'markdownfileid',
        class: 'text',
        mime: 'text/markdown',
        name: 'sample.md'
      },
      {
        _id: 'videofileid',
        class: 'video',
        mime: 'video/mp4',
        name: 'sample.mp4'
      },
      {
        _id: 'unknownfileid',
        class: '???',
        mime: 'something/unsupported',
        name: 'sample.unsupported'
      },
      {
        _id: 'shortcutfileid',
        class: 'shortcut',
        mime: 'application/internet-shortcut',
        name: 'cozy.url'
      }
    ]

    const component = shallow(
      <Viewer files={files} breakpoints={{}} currentIndex={0} />,
      shallowRenderOptions
    )

    const audioViewer = await getViewerComponent(component)
    expect(audioViewer).toMatchSnapshot('audioviewer')

    component.setProps({
      currentIndex: 1
    })
    const imageViewer = await getViewerComponent(component)
    expect(imageViewer).toMatchSnapshot('imageviewer')

    component.setProps({
      currentIndex: 2
    })
    const pdfViewer = await getViewerComponent(component)
    expect(pdfViewer).toMatchSnapshot('pdfviewer')

    component.setProps({
      currentIndex: 3
    })
    const textViewer = await getViewerComponent(component)
    expect(textViewer).toMatchSnapshot('textviewer')

    component.setProps({
      currentIndex: 4
    })
    const markdownViewer = await getViewerComponent(component)
    expect(markdownViewer).toMatchSnapshot('markdownviewer')

    component.setProps({
      currentIndex: 5
    })
    const videoViewer = await getViewerComponent(component)
    expect(videoViewer).toMatchSnapshot('videoviewer')

    component.setProps({
      currentIndex: 6
    })
    const noViewer = await getViewerComponent(component)
    expect(noViewer).toMatchSnapshot('noviewer')

    component.setProps({
      currentIndex: 7
    })
    const shortcutViewer = await getViewerComponent(component)
    expect(shortcutViewer).toMatchSnapshot('shortcutviewer')
  })
})
