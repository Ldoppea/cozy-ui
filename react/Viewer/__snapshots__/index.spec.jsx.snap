// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Viewer should render the various viewers: audioviewer 1`] = `
<View
  file={
    Object {
      "_id": "audiofileid",
      "class": "audio",
      "mime": "audio/mp3",
      "name": "sample.mp3",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: imageviewer 1`] = `
<View
  file={
    Object {
      "_id": "imagefileid",
      "class": "image",
      "mime": "image/jpg",
      "name": "sample.jpg",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: markdownviewer 1`] = `
<View
  file={
    Object {
      "_id": "markdownfileid",
      "class": "text",
      "mime": "text/markdown",
      "name": "sample.md",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: noviewer 1`] = `
<View
  file={
    Object {
      "_id": "unknownfileid",
      "class": "???",
      "mime": "something/unsupported",
      "name": "sample.unsupported",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: pdfviewer 1`] = `
<View
  file={
    Object {
      "_id": "pdffileid",
      "class": "pdf",
      "mime": "application/pdf",
      "name": "sample.pdf",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: shortcutviewer 1`] = `
<View
  file={
    Object {
      "_id": "shortcutfileid",
      "class": "shortcut",
      "mime": "application/internet-shortcut",
      "name": "cozy.url",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: textviewer 1`] = `
<View
  file={
    Object {
      "_id": "textfileid",
      "class": "text",
      "mime": "text/plain",
      "name": "sample.txt",
    }
  }
  onClose={[Function]}
/>
`;

exports[`Viewer should render the various viewers: videoviewer 1`] = `
<View
  file={
    Object {
      "_id": "videofileid",
      "class": "video",
      "mime": "video/mp4",
      "name": "sample.mp4",
    }
  }
  onClose={[Function]}
/>
`;
